### ISSUE TITLE
Nyxt's shell specification 

### ISSUE THREAD

**Issue by aadcg** [atlas-engineer/nyxt#2202]
[A shell is the outermost layer around an operating system](https://en.wikipedia.org/wiki/Shell_(computing)). Basically, they define the user-OS interaction.

There exist CLI, TUI and GUI shells, and these can be implemented in several programming languages.

Nyxt is a shell, as so is Emacs or the Linux console. Bash, zshell or fish are programming languages that serve the purpose of specific kind of shell.

Programming languages, operating systems and other kinds of critical software are often properly defined with a specification, from which implementations follow. Oddly enough, I'm not aware of any shell specification.  Shells seem to grow by trial and error.

After so many years of hackerdom, it's easy to have an idea of what defines a powerful shell. Arguably, Emacs is the most powerful shell today. There's hardly any doubt that a Lisp-like languages is the most suitable to write a shell. The grey areas concern, for instance, the minimal set of concepts to define a shell, and their definitions; where is the line between those who developer and use the shell; etc.

In Nyxt's development, these questions pop up often. For instance, "should we have major and minor modes, or does the notion of modes suffices?"; "how should we design our `hook`s?"; "is window management Nyxt's responsability or not?".

In short, I'm advocating that having an "Emacs-like" shell should be a library, implemented according to a specification. 

Writing a specification of this kind is hard, but it has its benefits over time. Without such a document, I'm afraid it will get harder and harder to make sense of things from the codebase alone.

**Comment by Ambrevar** [comment]
I'm all for pushing parts of Nyxt to libraries.  We now have nhooks, nfiles, and hopefully more to come!
I can foresee:

- nwindows
- nclass (coming soon©)
- nkeymap (it's almost ready, needs some polishing)
- nprompt (same)

Now to the question of specs: if the libraries come with extensive test suites, part of the specs would be embedded there (programmatically).  Writing human-readable specs is a tremendous effort and I'm not sure the Nyxt team has the manpower to pull this off at the moment... Maybe in the future?

**Comment by aartaka** [comment]
As a yet another crazy Nyxt-as-an-OS idea: there's [Ubuntu Web Remix](https://discourse.ubuntu.com/t/ubuntu-web-remix/19394) that basically is Ubuntu with Firefox as the desktop environment and webapps as the substitution for regular desktop apps. We can do the same with Nyxt, except that we need to:
- [ ] Make file-manager fully usable with all the operations (deleting, removing, copying, moving, symlink-ing) available from the prompt and possible `file://` URL pages.
- [ ] Make more applications for everyday things, like 
  - [ ] terminal (Lisp REPL with some improvements and syntactic sugar?), 
  - [ ] email reader ([nx-notmuch](https://github.com/igoralmeida/nx-notmuch)?), 
  - [ ] text editor (improved `editor-buffer`, [Lem integration](https://github.com/lem-project/lem/issues/543), [NeoVim integration](https://neovim.io/doc/user/api.html)?)
- [ ] Add support for HTML-based webapps and intuitive leyboard-friendly modes for the suggested ones.
- [ ] Pack it up as a Linyx distributive, much like Ubuntu Web Remix above (@Ambrevar says it's easy, but we still need to sourt out things from the list above).

**Comment by Ambrevar** [comment]
"Linyx" is a nice name for our new distribution! :D

**Comment by aadcg** [comment]
> Now to the question of specs: if the libraries come with extensive test suites, part of the specs would be embedded there (programmatically). Writing human-readable specs is a tremendous effort and I'm not sure the Nyxt team has the manpower to pull this off at the moment... Maybe in the future?

After some thought, I agree with you that it's not in our interests to go about it now. Still, I will personally invest a lot of time in proper documentation since it also helps us to make sense of the big picture.

We'd be the first to have a shell spec as far as I know.

**Comment by aadcg** [comment]
The file manager has also been referred to in #2264.

**Comment by Ambrevar** [comment]
You might be interested in what I've been doing here:

https://github.com/atlas-engineer/nyxt/pull/2263

This is essentially a "Common Lisp spec" done via defgeneric.

The main drawback I see for now is that function typing is poor.  Does anyone know a way to type generic functions more precisely?

**Comment by Ambrevar** [comment]
This won't do for optional and keyword arguments, also it won't do for the return type.

**Comment by aartaka** [comment]
Then we're left with `check-type` and `declare`ations in the method body ¯\\\_(ツ)\_/¯

**Comment by aartaka** [comment]
Yeah. Let's optimize once we hit the performance problem, it should be fine so far :)

**Comment by Ambrevar** [comment]
I'm not saying this for performance, but for the matter at hand: to describe de specs _programmatically_.
